ballista/core/src/execution_plans/shuffle_writer.rs

```diff
@@ -288,6 +305,22 @@ impl ShuffleWriterExec {
     }
 }
 
+#[async_trait]
+impl ShuffleWriter for ShuffleWriterExec {
```

Comment:
This seems inverted to me. Shouldn't `ShuffleWriterExec` have a `Arc<dyn ShuffleWriter>` instead of impl the trait? 

I know this is a step toward the pluggable execution engine, but also thinking about how this fits with allowing for different shuffle methods (write to disk, write to object store, stream to network, etc.) and seems like we would need a second layer of indirection for that.

Reply:
That's a good point. Maybe want I want for now is just an `ExecutionEngine` trait with an `execute` method that returns the metadata with shuffle file locations. I will give this some more thought.

Reply:
@thinkharderdev I have renamed the trait and made some other naming changes. I also added the ExecutionEngine abstraction in this PR. PTAL when you have time.